### PR TITLE
Implement the "drain" feature for subscriptions and connections

### PR DESCRIPTION
**Pull request by jsudano**
Based on the proposal in #1325, implementing the functionality described in [this NATS doc](https://docs.nats.io/using-nats/developer/receiving/drain). 

I would love suggestions for further tests. I tried adding tests for more complex timing/corner cases but they all wound up looking like exactly like the tests I have here

### CODE REVIEW

File: async-nats/src/lib.rs [hunk 1]
```diff
@@ -1251,6 +1293,48 @@ impl Subscriber {
             .await?;
         Ok(())
     }
+
+    /// Unsubscribes immediately but leaves the stream open to allow any in-flight messages on the
```
Comment (Jarema) [inline async-nats/src/lib.rs:1297]: Maybe `subscription` instead of `stream`?  Sounds more direct on what is affected.

File: async-nats/src/lib.rs [hunk 2]
```diff
@@ -1251,6 +1293,48 @@ impl Subscriber {
...
+    /// subscription to be delivered. The stream will be closed after any remaining messages are
+    /// delivered
+    ///
+    /// # Examples
+    /// ```
```
Comment (Jarema) [inline async-nats/src/lib.rs:1302]: Please add `no_run` to examples - the demo server is not reliable enough :).

**Review by Jarema — CHANGES_REQUESTED** [review]
Thank you for this contribution!

This looks really good. 
Some comments added.

**Comment by Jarema** [comment]
(CI was red because there was a typo in example. Should be fine after a rebase)

**Comment by Jarema** [comment]
@jsudano ping, if you missed the review 🙂

**Comment by jsudano** [comment]
Apologies for the delay, was away from the computer last week. Really appreciate the review, will get the comments addressed this week!

File: async-nats/src/lib.rs [hunk 3]
```diff
@@ -773,6 +797,26 @@ impl ConnectionHandler {
...
+                        .enqueue_write_op(&ClientOp::Unsubscribe { sid, max: None });
+                };
+
+                if let Some(sid) = sid {
+                    if let Some(sub) = self.subscriptions.get_mut(&sid) {
+                        drain_sub(sid, sub);
+                    }
+                } else {
+                    // sid isn't set, so drain the whole client
+                    self.connector.events_tx.try_send(Event::Draining).ok();
```
Comment (jsudano) [inline async-nats/src/lib.rs:813]: Assumed it was fine to ignore any errors here as they'd be pretty unlikely (the only scenario I could think of would be if the drain command was sent mid-shutdown, in which case it's fine to just ignore the error).

File: async-nats/src/lib.rs [hunk 4]
```diff
@@ -630,7 +650,11 @@ impl ConnectionHandler {
                     };
                     debug!("reconnected");
                 }
-                ExitReason::Closed => break,
+                ExitReason::Closed => {
+                    //
+                    self.connector.events_tx.try_send(Event::Closed).ok();
```
Comment (jsudano) [inline async-nats/src/lib.rs:655]: Similar to the other comment, this is more-or-less the last thing we'd do during shutdown so it's probably fine to ignore errors.

Comment (Jarema) [reply async-nats/src/lib.rs:655]: It is, however please either remove the `//` or add a comment there that explains why its ok.

**Review by Jarema — APPROVED** [review]
LGTM!

Just one leftover from I guess intended comment left.

**Review by Jarema — CHANGES_REQUESTED** [review]
Oh sorry, missed the cargo fmt and linter errors. 
After fixing them, it's good to merge ;).

**Comment by jsudano** [comment]
Oops, hopefully that last commit got all of them (I ran the commands from the validation runs locally and they passed at least)

**Comment by Jarema** [comment]
@jsudano Thanks!

The test is missing ` use futures_util::stream::stream::StreamExt;`

**Review by Jarema — APPROVED** [review]
LGTM!
